client/service/src/chain_ops.rs

```diff
@@ -116,55 +300,32 @@ impl<
 			let client = &self.client;
 			let queue = &mut self.import_queue;
 
-			// Start by reading the number of blocks if not done so already.
-			let count = match count {
-				Some(c) => c,
-				None => {
-					let c: u64 = match Decode::decode(&mut io_reader_input) {
-						Ok(c) => c,
-						Err(err) => {
-							let err = format!("Error reading file: {}", err);
-							return std::task::Poll::Ready(Err(From::from(err)));
-						},
-					};
-					info!("📦 Importing {} blocks", c);
-					count = Some(c);
-					c
-				}
-			};
-
 			// Read blocks from the input.
-			if read_block_count < count {
-				match SignedBlock::<Self::Block>::decode(&mut io_reader_input) {
-					Ok(signed) => {
-						let (header, extrinsics) = signed.block.deconstruct();
-						let hash = header.hash();
-						// import queue handles verification and importing it into the client
-						queue.import_blocks(BlockOrigin::File, vec![
-							IncomingBlock::<Self::Block> {
-								hash,
-								header: Some(header),
-								body: Some(extrinsics),
-								justification: signed.justification,
-								origin: None,
-								allow_missing_state: false,
-								import_existing: force,
-							}
-						]);
+			match Pin::new(&mut block_stream).poll_next(cx) {
+				Poll::Ready(None) => {
+					let read_block_count = block_stream.read_block_count();
+					let count = block_stream.count();
+
+					if importing_is_done(count, read_block_count, link.imported_blocks) {
+						info!(
+							"🎉 Imported {} blocks. Best: #{}",
+							read_block_count, client.chain_info().best_number
+						);
+						return std::task::Poll::Ready(Ok(()))
 					}
-					Err(e) => {
-						warn!("Error reading block data at {}: {}", read_block_count, e);
-						return std::task::Poll::Ready(Ok(()));
+				},
+				Poll::Ready(Some(block_result)) => {
+					match block_result {
+						Ok(signed_block) => import_block_to_queue(signed_block, queue, force),
+						Err(e) => {
+							let read_block_count = block_stream.read_block_count();
+							warn!("Error reading block data at {}: {}", read_block_count, e);
+							// We've encountered an error, we can stop here.
+							return std::task::Poll::Ready(Ok(()));
+						}
 					}
 				}
-
-				read_block_count += 1;
-				if read_block_count % 1000 == 0 {
-					info!("#{} blocks were added to the queue", read_block_count);
-				}
-
-				cx.waker().wake_by_ref();
-				return std::task::Poll::Pending;
+				Poll::Pending => unreachable!("BlockStream.poll_next() should never return Poll::Pending; qed")
```

Comment:
`BlockStream` can just implement `Iterator`, there is no reason it implements `Stream` as it is not async.

Reply:
Correct! I'll move this to `Iterator` and switch to stream in the next PR :)

Reply:
Why switch to `Stream` in the next pr?

Reply:
So I'm planning on re-writing the import block as an async block in a subsequent PR. For some reason I thought that I would need to use a Stream if I were to use an async block, but I guess I was wrong :)

Reply:
I think the `import_blocks` function was originally made as async so that's why you used Stream which makes sense.

But when I read the code I saw that it was quite artificial, it's just a big loop splitted in chunks to do the import so other async code may run in parallel.

I believe this async-ness was not necessary at all because the process is not parallelizing the import and this import-blocks method is only run on its own with the subcommand `import-blocks`. So it feels to me that the original design was a bit overengineered and it can be a simple blocking code. But I might be missing something important and I might be wrong.

Reply:
Just moved from Stream to Iterator in https://github.com/paritytech/substrate/pull/5816/commits/994157c87d4c518ba5023b232acbb5e6b71f9635 :D

Reply:
Ahhh you meant literally an iterator XD is that simpler? I thought the code could be converted to a `loop {}` block but just not async.

Reply:
Hmm not sure I follow haha ^^
You're talking about completely removing the async?

Reply:
Yes! I might be wrong but this code doesn't really looks async, I don't see it awaiting anything anywhere

Reply:
Maybe @bkchr would have an idea here? I must admit I'm still not 100% comfortable with async stuff :p

Reply:
> Yes! I might be wrong but this code doesn't really looks async, I don't see it awaiting anything anywhere

But we are returning some `Poll`s... I think having an iterator is a pretty clean approach :D

Reply:
Nevermind for now, probably it's doing something async I just don't see it. It will become clearer when you will do the next PR that re-write this